src/_xoConnection.js

```diff
@@ -0,0 +1,181 @@
+      forOwn(params.items, (item, id) => {
+        fn.call(objects, id, item);
+
+        const watcher = watchers[id];
+        if (watcher !== undefined) {
+          watcher(item);
+          delete watchers[id];
+        }
+      });
+    });
+  }
+
+  async _fetchObjects() {
+    const { _objects: objects, _watchers: watchers } = this;
+    forOwn(await this.call("xo.getAllObjects"), (object, id) => {
+      objects.set(id, object);
+
+      const watcher = watchers[id];
+      if (watcher !== undefined) {
+        watcher(object);
+        delete watchers[id];
+      }
+    });
+  }
+
+  // TODO: integrate in xo-lib.
+  waitObject(id) {
+    return new Promise(resolve => {
+      this._watchers[id] = resolve;
+    }); // FIXME: work with multiple listeners.
+  }
+
+  async getOrWaitObject(id) {
+    const object = this._objects.all[id];
+    if (object !== undefined) {
+      return object;
+    }
+    return this.waitObject(id);
+  }
+
+  @defer
+  async connect($defer, credentials = config.adminCredentials) {
+    await this.open();
+    $defer.onFailure(() => this.close());
+    await this.signIn(credentials);
+
+    this._addObjectsListener();
+    await this._fetchObjects();
+
+    return this;
+  }
+
+  async waitObjectState(id, predicate) {
+    let obj = this._objects.all[id];
+    while (true) {
+      try {
+        await predicate(obj);
+        return;
+      } catch (_) {}
+      // If failed, wait for next object state/update and retry.
+      obj = await this.waitObject(id);
+    }
+  }
+
+  async _createTempResources(method, params) {
+    const result = await super.call(method, params);
+    const [class_] = method.split(".");
+    this._tempResourcesByClass[class_].push(result.id || result);
+    return result;
+  }
+
+  async _deleteTempResources(method, params) {
+    const [class_] = method.split(".");
+    const tempResources = this._tempResourcesByClass[class_];
+    await Promise.all(
+      tempResources.map(id =>
+        super.call(method, { id, ...params }).catch(logError)
+      )
+    );
+    tempResources.length = 0;
+  }
+
+  createUser(params) {
+    return this._createTempResources("user.create", params);
+  }
+
+  deleteAllUsers() {
+    return this._deleteTempResources("user.delete");
+  }
+
+  async getUser(id) {
+    return find(await super.call("user.getAll"), { id });
+  }
+
+  createTempJob(params) {
+    return this._createTempResources("job.create", { job: params });
+  }
+
+  deleteTempJobs() {
+    return this._deleteTempResources("job.delete");
+  }
+
+  createTempBackupNgJob(params) {
+    return this._createTempResources("backupNg.createJob", params);
+  }
+
+  deleteTempBackupNgJobs() {
+    return this._deleteTempResources("backupNg.deleteJob");
+  }
+
+  deleteTempResources() {
+    return Promise.all([
+      this.deleteAllUsers().catch(logError),
+      this.deleteTempJobs().catch(logError),
+      this.deleteTempBackupNgJobs().catch(logError),
```

Comment:
What's the point of exposing these methods?

Are they used outside `deleteTempResources`?
Would it make sense to use them?

Reply:
No, i'll delete them :+1: